Use autoconf to generate the compiler's configuration script

shindere opened the pull request:
Cc @pgraverdy

## Introduction

This GPR continues the work started in GPRs #2044, #2059, #2113 and #2115. At a higher level: after having merged the Unix and Windows build systems, this PR merges their configuration systems.

To do so, the handwritten configuration script is replaced by an autoconf-generated one, which is also expected to make the future work on producing cross-compilers easier. It becomes possible to configure the Windows builds similarly to what can be done under Unix and without having to tweak the makefiles. (This GPR does start to update README files, but the update should be completed in a future GPR.)

## Command-line options to the configure script

They change completely because autoconf imposes restrictions on their form. The new options are however documented in `./configure --help`, which was not the case previously.

Still, for people packaging the OCaml compiler, it may be useful to provide a mapping between the old and new option syntax. I would be willing to do that if it is deemed necessary, but I am not sure where it would best fit. It could be done in a distinct GPR that could even be merged after the feature freeze.

## libtool

Basically, autoconf by itself does not provide any means of figuring out anything related to libraries (where the linker is, which options to use to build shared libraries, etc.). To do that, one is expected to rely on libtool, which provides its own set of macros that should be invoked in the configuration script. These macros generate a `libtool` script that one is supposed to use during the build to compile and link files, rather than calling the compiler and linker directly.

The approach implemented in this GPR does use the `libtool` configuration-time machinery to figure things out but at the same time does not call `libtool` to compile and link files. This approach has been chosen first because introducing `libtool` in the build system seemed a too big change, second because on Windows some libtool guesses were wrong and it was also not clear how well libtool would interact with flexlink.

(Now that we have libtool, nothing prevents us from relying on it more in the future and, should we want to do so, it will be possible to do the transition in an incremental way.)

An important consequence of all this is that some of the guesses shown by configure (such as the linker location) will actually not be used. This is because, once the libtool-related detection macros have been invoked, some of its results are ignored and replaced by the ones that were previously hardcoded.

This may look awkward and, to some extent, it certainly is, but at the same time it seemed that this strategy makes the transition to autoconf smooth by avoiding to modify the rest of the build system as much as possible.

## Regressions

This GPR introduces two regressions, to be fixed before the 4.08 final release:

 1. The former `-target-bind-dir` option has not been implemented yet;
 2. The `unwind` library is not yet detected.

Due to time constraints, it seemed better to submit the code in its current state to give people more time to test and discuss. Adding the support for these two features can be done in parallel with the discussion.

shindere committed:
config/Makefile.m*: remove the no longer used BYTECCDBGCOMPOPTS variable

dra27 commented:
Bravo @shindere! Review and tests at my end to follow

shindere commented:
Note to reviewers: all the files in the `build-aux` directory come from either autoconf or libtool -- it is a standard practice to bundle them. Of course, `configure` is generated. So all these files do normally not need to be reviewed.

shindere commented:
Many thanks, @dra27! Beware, the code in `configure.ac` is very far from
being perfect, so let's just make sure it wokrks so that we have a basis
on which we can start to improve and cleanup things step by step.

shindere committed:
Use autoconf to generate the compiler's configuration script

shindere commented:
@gasche: one Travis-related question, so I am assuming that's for you.

Why is flambda enabled in .travis.yml, whereas flambda-invariants is
enabled in the shell script? Do you know / remember?

damiendoligez commented on aclocal.m4:
```diff
@@ -0,0 +1,254 @@
+clang __clang_major__ __clang_minor__
+#elif defined(__GNUC__) && defined(__GNUC_MINOR__)
+gcc __GNUC__ __GNUC_MINOR__
+#elif defined(__xlc__) && (__xlC__)
```
Shouldn't this be `&& defined (__xlC__)` ?

damiendoligez commented on README.win32.adoc:
```diff
@@ -196,16 +196,11 @@ quickly as it will be unable to link `ocamlrun`.
 for 32-bit, or:
 
-        cp config/Makefile.msvc64 Makefile.config
+        ./configure --build=x86_64-unknown-cygwin --host=x86_64-pc-windows
```
I'm curious to know why it's `unknown` here but `pc` on i686...

damiendoligez commented on config/Makefile.mingw:
```diff
@@ -118,8 +118,6 @@ OC_CPPFLAGS=-DCAML_NAME_SPACE -DUNICODE -D_UNICODE \
 OCAMLC_CFLAGS=-O -mms-bitfields
 
-BYTECCDBGCOMPOPTS=-g
-
```
Didn't this file become obsolete? (as well as the other `Makefile.m*` files)

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+
+AC_ARG_WITH([libunwind-include-dir],
+  [AS_HELP_STRING([--with-libunwindlib],
+    [location of binary files for libunwind])])
```
`include-dir` is for header files, not binary files.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+
+AC_ARG_ENABLE([frame-pointers],
+  [AS_HELP_STRING([--enable-frame-pointers],
+    [build the runtime system with frame pointers])])
```
Not just the runtime system: the `frame-pointers` option also changes the code generated by the compiler.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+
+AC_ARG_ENABLE([naked-pointers],
+  [AS_HELP_STRING([--disable-naked-pointers],
+    [do not use naked pointers])])
```
Should be `do not allow` rather than `do not use`.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+# explicitly passed.
+#
+# The configure-time behavior of OCaml 4.05 and older was equivalent
+# to --disable-force-safe-string --with-default-string=unsafe. OCaml 4.06
```
Ocaml 4.06 and later.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+
+AC_ARG_ENABLE([force-safe-string],
+  [AS_HELP_STRING([--disable-force-safe-string],
+    [do not provide the -force-safe-string option])])
```
Do not force strings to be safe.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+
+AC_ARG_WITH([default-string],
+  [AS_HELP_STRING([--with-default-string=mode],
+    [wether strings should be safe or unsafe @<:@default=safe@:>@])])
```
You should add "by default" somewhere in this sentence.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+  # ARCMD=$(TOOLPREF)AR
+  # RANLIB=$(TOOLPREF)ranlib
+  # RANLIBCMD=$(TOOLPREF)ranlib
+  # However autoconf and libtool seem to use ar and ranlib
```
I don't understand this comment. What you describe above also uses ar and ranlib, so what's the difference? Capitalization?

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+natdynlink=false
+
+AS_IF([test x"$enable_shared" != "no"],
+  [AS_CASE(["$host"],
```
Why is this test on `$host` ? In the old `configure` script it's on `$target`.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+AS_IF([test x"$enable_shared" != "no"],
+  [AS_CASE(["$host"],
+    [*-*-cygwin*], [natdynlink=true],
+    [*-*-mingw32], [natdynlink=true],
```
In the original configure, it's on `mingw*`. Did you change it on purpose?

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+model=default
+system=unknown
+
+AS_CASE([$host],
```
Same question with host/target.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+    [AS_IF([$arch64], [arch=amd64], [arch=i386]); system=macosx],
+  [[i[3456]86-*-gnu*]],
+    [arch=i386; system=gnu],
+  [[i[3456]86-*-mingw32]],
```
same question with mingw32/mingw*

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+    [arch=amd64; system=openbsd],
+  [x86_64-*-darwin*],
+    [arch=amd64; system=macosx],
+  [x86_64-*-mingw32],
```
Same question again with mingw32/mingw*
It seems more important here, since you're turning mingw32 into mingw64.

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+    [power,gcc*,elf,ppc64], [PARTIALLD="ld -r -m elf64ppc"],
+    [power,gcc*,elf,ppc64le], [PARTIALLD="ld -r -m elf64lppc"],
+    [*,cl,*,*], [PARTIALLD="link -lib -nologo"],
+    [PARTIALLD="ld -r"])])
```
You removed the command-line option to override partialld?

damiendoligez commented on tools/ci/travis/travis-ci.sh:
```diff
@@ -80,15 +80,18 @@ critical errors that must be understood and fixed before your pull
   x64)
-    ./configure --prefix $PREFIX -with-debug-runtime \
-      -with-instrumented-runtime -with-flambda-invariants $CONFIG_ARG
+    ./configure $configure_flags
```
I guess the debug runtime is on by default, but you removed `-with-instrumented-runtime` ??

damiendoligez commented on tools/ci/travis/travis-ci.sh:
```diff
@@ -80,15 +80,18 @@ critical errors that must be understood and fixed before your pull
-      -with-instrumented-runtime -with-flambda-invariants $CONFIG_ARG \
-      -host i686-pc-linux-gnu
+    ./configure --build=x86_64-pc-linux-gnu --host=i386-pc-linux-gnu \
+      $configure_flags
```
same remark

damiendoligez commented on configure.ac:
```diff
@@ -0,0 +1,1657 @@
+ocamlopt_cflags=""
+ocamlopt_cppflags=""
+oc_ldflags=""
+wish_sharedlibs=true
```
typo `with_sharedlibs` (bug)

damiendoligez reviewed (commented):
A few remarks from having a quick look.

shindere commented:
Thanks a lot for the review and sorry for the delay for taking it into
account.

Damien Doligez (2018/11/07 07:55 -0800):
> > +# https://sourceforge.net/p/predef/wiki/Compilers/
> +AC_DEFUN([OCAML_CC_VENDOR], [
> +  AC_REQUIRE([AC_PROG_CC])
> +  AC_REQUIRE([AC_PROG_CPP])
> +  AC_MSG_CHECKING([C compiler vendor])
> +  AC_PREPROC_IFELSE(
> +    [AC_LANG_SOURCE([
> +#if defined(_MSC_VER)
> +msvc _MSC_VER
> +#elif defined(__INTEL_COMPILER)
> +icc __INTEL_COMPILER
> +#elif defined(__clang_major__) && defined(__clang_minor__)
> +clang __clang_major__ __clang_minor__
> +#elif defined(__GNUC__) && defined(__GNUC_MINOR__)
> +gcc __GNUC__ __GNUC_MINOR__
> +#elif defined(__xlc__) && (__xlC__)
> 
> Shouldn't this be `&& defined (__xlC__)` ?

Well that line comes from `config/auto-aux/cckind.c`, see line 28. You
may be right but (1) it's hard to say without being able to test with
the xlc compiler and (2) one could also argue that the purpose of this
PR is to switch to autoconf and not to fix bugs (I agree that would be a
very bad way of reasoning).

>  for 32-bit, or:
>  
> -        cp config/Makefile.msvc64 Makefile.config
> +        ./configure --build=x86_64-unknown-cygwin --host=x86_64-pc-windows
> 
> I'm curious to know why it's `unknown` here but `pc` on i686...

I don't know. It's just the values guessed / recognized by
config.guess...

> > @@ -118,8 +118,6 @@ OC_CPPFLAGS=-DCAML_NAME_SPACE -DUNICODE -D_UNICODE \
>    -DWINDOWS_UNICODE=$(WINDOWS_UNICODE)
>  OCAMLC_CFLAGS=-O -mms-bitfields
>  
> -BYTECCDBGCOMPOPTS=-g
> -
> 
> Didn't this file become obsolete? (as well as the other `Makefile.m*`
> files)

Well actually, it's the whole config/ subdirectory that would become
obsolete but I didn't dare to remove it right now. But I can, if you so
wish. It's also why I insisted on moving `config/Makefile` to
`Makefile.config`, because otherwise we would have had to keep the
config directory only for that file.

> > +  [AS_HELP_STRING([--enable-instrumented-runtime],
> +    [build the instrumented runtime @<:@default=auto@:>@])],
> +  [],
> +  [enable_instrumented_runtime=auto])
> +
> +AC_ARG_ENABLE([pthread],
> +  [AS_HELP_STRING([--disable-pthread],
> +    [disable the POSIX thread library])])
> +
> +AC_ARG_WITH([libunwind],
> +  [AS_HELP_STRING([--with-libunwind],
> +    [location of the libunwind package])])
> +
> +AC_ARG_WITH([libunwind-include-dir],
> +  [AS_HELP_STRING([--with-libunwindlib],
> +    [location of binary files for libunwind])])
> 
> `include-dir` is for header files, not binary files.
> 
> > +
> +AC_ARG_WITH([libunwind-include-dir],
> +  [AS_HELP_STRING([--with-libunwindlib],
> +    [location of binary files for libunwind])])
> +
> +AC_ARG_ENABLE([graph-lib],
> +  [AS_HELP_STRING([--disable-graph-lib],
> +    [do not build the graphics library])])
> +
> +AC_ARG_ENABLE([ocamldoc],
> +  [AS_HELP_STRING([--disable-ocamldoc],
> +    [do not build the ocamldoc documentation system])])
> +
> +AC_ARG_ENABLE([frame-pointers],
> +  [AS_HELP_STRING([--enable-frame-pointers],
> +    [build the runtime system with frame pointers])])
> 
> Not just the runtime system: the `frame-pointers` option also changes
> the code generated by the compiler.

The help string is now "use frame pointers in runtime and generated code"
 
> +AC_ARG_ENABLE([graph-lib],
> +  [AS_HELP_STRING([--disable-graph-lib],
> +    [do not build the graphics library])])
> +
> +AC_ARG_ENABLE([ocamldoc],
> +  [AS_HELP_STRING([--disable-ocamldoc],
> +    [do not build the ocamldoc documentation system])])
> +
> +AC_ARG_ENABLE([frame-pointers],
> +  [AS_HELP_STRING([--enable-frame-pointers],
> +    [build the runtime system with frame pointers])])
> +
> +AC_ARG_ENABLE([naked-pointers],
> +  [AS_HELP_STRING([--disable-naked-pointers],
> +    [do not use naked pointers])])
> 
> Should be `do not allow` rather than `do not use`.

Fixed, thanks
 
> > +#
> +# If --enable-force-safe-string is set at configure time, then the compiler
> +# will always enforce that string and bytes are distinct: the
> +# compile-time -unsafe-string option is disabled. This lets us
> +# assume pervasive string immutability, for code optimizations and
> +# in the C layer.
> +#
> +# If --disable-force-safe-string is set at configure-time, the compiler
> +# will use the compile-time (un)safe-string option to decide whether
> +# string and bytes are compatible on a per-file basis. The
> +# configure-time options --with-default-string=safe|unsafe decides which
> +# setting will be chosen by default, if no compile-time option is
> +# explicitly passed.
> +#
> +# The configure-time behavior of OCaml 4.05 and older was equivalent
> +# to --disable-force-safe-string --with-default-string=unsafe. OCaml 4.06
> 
> Ocaml 4.06 and later.

Fixed, thanks. Was copied from the existing configure script.

> +# If --disable-force-safe-string is set at configure-time, the compiler
> +# will use the compile-time (un)safe-string option to decide whether
> +# string and bytes are compatible on a per-file basis. The
> +# configure-time options --with-default-string=safe|unsafe decides which
> +# setting will be chosen by default, if no compile-time option is
> +# explicitly passed.
> +#
> +# The configure-time behavior of OCaml 4.05 and older was equivalent
> +# to --disable-force-safe-string --with-default-string=unsafe. OCaml 4.06
> +# uses --disable-force-safe-string --with-default-string=safe. We
> +# expect --enable-force-safe-string to become the default in the future.
> +
> +AC_ARG_ENABLE([force-safe-string],
> +  [AS_HELP_STRING([--disable-force-safe-string],
> +    [do not provide the -force-safe-string option])])
> 
> Do not force strings to be safe.

Fixed, thanks

> > +# configure-time options --with-default-string=safe|unsafe decides which
> +# setting will be chosen by default, if no compile-time option is
> +# explicitly passed.
> +#
> +# The configure-time behavior of OCaml 4.05 and older was equivalent
> +# to --disable-force-safe-string --with-default-string=unsafe. OCaml 4.06
> +# uses --disable-force-safe-string --with-default-string=safe. We
> +# expect --enable-force-safe-string to become the default in the future.
> +
> +AC_ARG_ENABLE([force-safe-string],
> +  [AS_HELP_STRING([--disable-force-safe-string],
> +    [do not provide the -force-safe-string option])])
> +
> +AC_ARG_WITH([default-string],
> +  [AS_HELP_STRING([--with-default-string=mode],
> +    [wether strings should be safe or unsafe @<:@default=safe@:>@])])
> 
> You should add "by default" somewhere in this sentence.

Done, thanks

> > +LT_INIT
> +
> +# Extracting information from libtool's configuration
> +AS_IF([test -n "$RANLIB" ],
> +  [RANLIBCMD="$RANLIB"],
> +  [RANLIB="$AR rs"; RANLIBCMD=""]
> +)
> +
> +AS_CASE([$host],
> +  # In config/Makefile.mingw*, we had:
> +  # TARGET=i686-w64-mingw32 and x86_64-w64-mingw32
> +  # TOOLPREF=$(TARGET)-
> +  # ARCMD=$(TOOLPREF)AR
> +  # RANLIB=$(TOOLPREF)ranlib
> +  # RANLIBCMD=$(TOOLPREF)ranlib
> +  # However autoconf and libtool seem to use ar and ranlib
> 
> I don't understand this comment. What you describe above also uses ar
> and ranlib, so what's the difference? Capitalization?

No the difference is that what is found by autoconf does not have a
prefix.

> > +  [*-*-mingw32],
> +    [mksharedlib='$(FLEXLINK)'
> +    mkmaindll='$(FLEXLINK) -maindll'
> +    shared_libraries_supported=true],
> +  [*-pc-windows],
> +    [mksharedlib='$(FLEXLINK)'
> +    mkmaindll='$(FLEXLINK) -maindll'])
> +
> +AS_IF([test -z "$mkmaindll"], [mkmaindll=$mksharedlib])
> +
> +# Configure native dynlink
> +
> +natdynlink=false
> +
> +AS_IF([test x"$enable_shared" != "no"],
> +  [AS_CASE(["$host"],
> 
> Why is this test on `$host` ? In the old `configure` script it's on
> `$target`.

Well I think the legacy configure script had some confusion between host
and target. This will certainly have to be clarified when the th ework
on cross-compiling starts, but at the moment it seemed reasonable to do
most of the tests on host...

> > +    mkmaindll='$(FLEXLINK) -maindll'
> +    shared_libraries_supported=true],
> +  [*-pc-windows],
> +    [mksharedlib='$(FLEXLINK)'
> +    mkmaindll='$(FLEXLINK) -maindll'])
> +
> +AS_IF([test -z "$mkmaindll"], [mkmaindll=$mksharedlib])
> +
> +# Configure native dynlink
> +
> +natdynlink=false
> +
> +AS_IF([test x"$enable_shared" != "no"],
> +  [AS_CASE(["$host"],
> +    [*-*-cygwin*], [natdynlink=true],
> +    [*-*-mingw32], [natdynlink=true],
> 
> In the original configure, it's on `mingw*`. Did you change it on
> purpose?

Beware that the mingw parts in the old configure are dead code. The real
reference for MinGW is in the makefiles under config/

> > +    [aarch64-*-linux*], [natdynlink=true])])
> +
> +# Try to work around the Skylake/Kaby Lake processor bug.
> +AS_CASE(["$CC,$host"],
> +  [*gcc*,x86_64-*|*gcc*,i686-*], 
> +    [OCAML_CC_HAS_FNO_TREE_VRP
> +    AS_IF([$cc_has_fno_tree_vrp],
> +      [internal_cflags="$internal_cflags -fno-tree-vrp"])])
> +
> +# Configure the native-code compiler
> +
> +arch=none
> +model=default
> +system=unknown
> +
> +AS_CASE([$host],
> 
> Same question with host/target.

Same response. :)

> > +system=unknown
> +
> +AS_CASE([$host],
> +  [[i[3456]86-*-linux*]],
> +    [arch=i386; system=linux_elf],
> +  [[i[3456]86-*-*bsd*]],
> +    [arch=i386; system=bsd_elf],
> +  [[i[3456]86-*-haiku*]],
> +    [arch=i386; system=beos],
> +  [[i[3456]86-*-cygwin]],
> +    [arch=i386; system=cygwin],
> +  [[i[3456]86-*-darwin*]],
> +    [AS_IF([$arch64], [arch=amd64], [arch=i386]); system=macosx],
> +  [[i[3456]86-*-gnu*]],
> +    [arch=i386; system=gnu],
> +  [[i[3456]86-*-mingw32]],
> 
> same question with mingw32/mingw*

Same response

> > +    [arch=arm; system=bsd],
> +  [x86_64-*-linux*],
> +    [arch=amd64; system=linux],
> +  [x86_64-*-gnu*],
> +    [arch=amd64; system=gnu],
> +  [x86_64-*-dragonfly*],
> +    [arch=amd64; system=dragonfly],
> +  [x86_64-*-freebsd*],
> +    [arch=amd64; system=freebsd],
> +  [x86_64-*-netbsd*],
> +    [arch=amd64; system=netbsd],
> +  [x86_64-*-openbsd*],
> +    [arch=amd64; system=openbsd],
> +  [x86_64-*-darwin*],
> +    [arch=amd64; system=macosx],
> +  [x86_64-*-mingw32],
> 
> Same question again with mingw32/mingw*

idem

> It seems more important here, since you're turning mingw32 into
> mingw64.

Which lines are we talking about in the freshly rebased code, please?

> > +
> +AS_IF([! $native_compiler], [natdynlink=false])
> +
> +AS_IF([$natdynlink], [cmxs="cmxs"], [cmxs="cmx"])
> +
> +AC_DEFINE_UNQUOTED([OCAML_OS_TYPE], ["$ostype"])
> +
> +AS_IF([test -z "$PARTIALLD"],
> +  [AS_CASE(["$arch,$CC,$system,$model"], 
> +    [amd64,gcc*,macosx,*], [PARTIALLD="ld -r -arch x86_64"],
> +    [amd64,gcc*,solaris,*], [PARTIALLD="ld -r -m elf_x86_64"],
> +    [power,gcc*,elf,ppc], [PARTIALLD="ld -r -m elf32ppclinux"],
> +    [power,gcc*,elf,ppc64], [PARTIALLD="ld -r -m elf64ppc"],
> +    [power,gcc*,elf,ppc64le], [PARTIALLD="ld -r -m elf64lppc"],
> +    [*,cl,*,*], [PARTIALLD="link -lib -nologo"],
> +    [PARTIALLD="ld -r"])])
> 
> You removed the command-line option to override partialld?

I may have overlooked that, sorry. But is it really useful somewhere?

> >    case $XARCH in
>    x64)
> -    ./configure --prefix $PREFIX -with-debug-runtime \
> -      -with-instrumented-runtime -with-flambda-invariants $CONFIG_ARG
> +    ./configure $configure_flags
> 
> I guess the debug runtime is on by default, but you removed
> `-with-instrumented-runtime` ??

It's called --enable/--disable-instrumented-runtime, now. By default the
instrumented runtime is built if the clock_gettime system call is
available, except on Windows where it's disabled explicitly by default
for compatibiilty reasons.

> > +
> +# Command to build executalbes
> +mkexe="\$(CC) \$(OC_CFLAGS) \$(OC_CPPFLAGS) \$(OC_LDFLAGS)"
> +
> +# Flags for building executable files with debugging symbols
> +mkexedebugflag="-g"
> +common_cflags=""
> +common_cppflags=""
> +internal_cflags=""
> +internal_cppflags=""
> +ocamlc_cflags=""
> +ocamlc_cppflags=""
> +ocamlopt_cflags=""
> +ocamlopt_cppflags=""
> +oc_ldflags=""
> +wish_sharedlibs=true
> 
> typo `with_sharedlibs` (bug)

Oops, fixed, thanks.

damiendoligez commented:
@shindere 
>Well actually, it's the whole config/ subdirectory that would become
obsolete but I didn't dare to remove it right now.

Sorry, my question was not clear. Why do you patch `config/Makefile.mingw` if it's now obsolete?

shindere commented:
Damien Doligez (2018/11/23 08:00 -0800):
> @shindere 
> >Well actually, it's the whole config/ subdirectory that would become
> obsolete but I didn't dare to remove it right now.
> 
> Sorry, my question was not clear. Why do you patch
> `config/Makefile.mingw` if it's now obsolete?

Because I wanted to make clear that the variable I removed was useless
and that this removal was independent of the switch to autoconf. In a
way it could even have been part of another GPR.

Am I making sense, now?

damiendoligez commented:
yes that makes sense

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+
+AC_ARG_ENABLE([pthread],
+  [AS_HELP_STRING([--disable-pthread],
+    [disable the POSIX thread library])])
```
I think this should be named `systhreads`, since the Windows implementation has nothing to do with POSIX:
```
AC_ARG_ENABLE([systhreads-lib], 
  [AS_HELP_STRING([--disable-systhreads-lib], 
    [disable the native threads (POSIX/Win32) library])])
```

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+
+AC_ARG_WITH([libunwind-include-dir],
+  [AS_HELP_STRING([--with-libunwind-include-dir],
+    [location of header files for libunwind])])
```
The unwind stuff seems to have several problems - there are large parts of the old `configure` (especially with macOS options) which don't appear to have been translated - is that intentional? Most importantly, the default seems to have changed - libunwind is meant to be required unless it's explicitly disabled? I know there was a discussion on caml-devel, but I wonder if the proper interface for this should be `--with-libunwind` with `LIBUNWIND_PREFIX` (default `""`), `LIBUNWIND_INCLUDE` (default is either `""` or `$LIBUNWIND_PREFIX/include` if `LIBUNWIND_PREFIX` is defined) and `LIBUNWIND_LIB` (same system for the default)

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+
+AC_ARG_ENABLE([call-counts],
+  [AS_HELP_STRING([--disable-call-counts],
+    [disable the call counts in spacetime])])
```
`enable-call-counts` only makes sense with `--enable-spacetime`. I therefore wonder if it would be better to have `--spacetime=mode` where `mode` is `with-call-counts` (the default) or `without-call-counts`

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+    [install *.cmt* and *.mli files])])
+AC_ARG_ENABLE([installing-bytecode-programs],
+  [AS_HELP_STRING([--enable-installing-bytecode-programs],
+    [also install the bytecode versions of programs])])
```
As a matter of personal taste, I find these options unusually named. Can I propose `--enable-source-install` and `--enable-bytecode-install` - it's the `installing` which seems very strange.

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+
+AC_ARG_ENABLE([native-compiler],
+  [AS_HELP_STRING([--disable-native-compiler],
+    [do not build the native compiler])])
```
Given again that it's changing name, can we consider `--enable-ocamlopt`?

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+
+AC_ARG_ENABLE([flambda-invariants],
+  [AS_HELP_STRING([--enable-flambda-invariants],
+    [enable invariants checks in flambda])])
```
As with `--enable-spacetime`, this is related to its parent option, so would `--flambda=with-invariants` be more logical?

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+
+AC_ARG_WITH([target-bindir],
+  [AS_HELP_STRING([--with-target-bindir],
+    [location of binary programs on target system])])
```
This parameter isn't propagated anywhere - it should be in `Makefile.config.in`. However, given how it's used, it should really be `AC_ARG_VAR([TARGET_BINDIR], ...)`

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+
+AC_ARG_WITH([reserved-header-bits],
+  [AS_HELP_STRING([--with-reserved-header-bits=BITS],
+    [reserve BITS bits in block headers for profiling info])])
```
This option hasn't quite preserved the semantics of the old one, although I'm not sure how relevant it is. Specifying `--with-reserved-header-bits` is also supposed to enable profinfo (even without spacetime). I wonder if it would be better to have `AC_ARG_ENABLE([profinfo], ...)` and have this as its parameter?

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+AC_ARG_WITH([default-string],
+  [AS_HELP_STRING([--with-default-string=mode],
+    [whether strings should be safe or unsafe by default
+      @<:@default=safe@:>@])])
```
This was a problem in the previous configure script, so perhaps now is the time to fix this properly. `--enable-force-safe-string --with-default-string=unsafe` is a conflicting configuration (I think it builds a compiler which displays an error message by default, but I haven't checked). I think this should one option `--enable-safe-string=mode` where `mode` is `no` (equivalent to `--disable-force-safe-string --with-default-string=unsafe`), `yes`/`default` (equivalent to `--disable-force-safe-string --with-default-string=safe`) or `always` (equivalent to `--enable-force-safe-string --with-default-string=safe`)

damiendoligez replied:
In trunk, `-force-safe-string` takes precedence. (see lines 167-170 of `utils/clflags.ml`)

damiendoligez replied:
I'd use `unsafe / safe / force` instead of `no / yes / always`.

gasche replied:
There is a conceptual difference between:
- `--force-safe-string`, which must be a configure-time option that changes the runtime definitions and has a global impact on all modules compiled and linked by the resulting compiler
- `--default-string={safe,unsafe}` which is "merely" a convenience feature to set a default for compile-time options, rather than a configure-time option by nature (`afl` is the same, etc.). It does nothing more than change the default for a choice that the user makes on a per-invocation basis.

gasche replied:
In my opinion this conceptual difference justifies *not* showing the three options as a single knob to fix at configure-time, but two separate kind of options. There is a dependency between them (`--force-safe-string` makes it incorrect to change the per-invocation value, so it makes `--default-string` useless and meaningless), but that is common across the configuration space, for example `--disable-native-compiler` will make many other settings meaningless.

dra27 commented on configure.ac:
```diff
@@ -0,0 +1,1666 @@
+  [AS_HELP_STRING([--with-dllibs=LIBS],
+    [use LIBS (in addition to dl) to load dynamic libs @<:@default=""@:>@])],
+  [otherdllibs=$withval],
+  [otherdllibs=""])
```
This should be `AC_ARG_VAR([DLLIBS], ...)`

dra27 reviewed (commented):
I've gone through the mapping of options to the old `configure` script and where they appear in the new one. Reading the various guides for autoconf, it seems to me that `AC_ARG_WITH` is intended to be used for external dependency related things only and so should only relate to the libunwind and afl options. That seems a more important rule to regard than the nonsense in the autoconf manual about `AC_ARG_ENABLE` not altering any features (since they don't provide any other kind of parameter for things like flambda, which is a feature which alters other features).

The old `configure` made no use of environment variables, so there are a few arguments which really should change to `AC_ARG_VAR` now, especially when they relate to passing directories in. Users have been used to saying `./configure -cc cc` so hopefully will switch to the correct `./configure CC=cc` and not incorrect `CC=cc ./configure`.

- The old `-lib*` parameter has no equivalent in the new one - should `$LIBS` be being put into `$cclibs`?
- ~~This is more of a todo for me, but a reminder - a patch of mine suggested changing `PARTIALLD` to `PACKLD` in `configure`. This is a mistake which I'll change - `configure` should accept `PARTIALLD` but `Makefile.config` should only contain `PACKLD`.~~ (this is fixed)
- The `-dldefs` parameter has no equivalent - as with `LIBS`, shouldn't `CFLAGS` be read at some point?
- `--with-cplugins` should be added, solely to maintain the error message.

Every single argument has changed, apart from `--prefix`, `--bindir`, `--libdir`, `--mandir`, `--host`, `--target` and `--verbose`. In doing this review, I now have a document which maps the old options to what should be their new equivalents. I think we should either insert a one-time script in front of `configure.ac` which translates the parameters (and which would never be updated with any new ones, obviously) or update the documentation with this mapping. My preference is for the script, which I'm happy to write, since it maintains as smooth a path as possible.

I don't think `--disable-shared` is working properly, but it's awkward testing both this and #2160 simultaneously, so I'll come back to that after this GPR is merged!

shindere commented:
Thanks for your feedback, David. Will go back to it tomorrow.

Meanwhile: did you read the text written to introduce this PR? I think
it may respond to a few of your comments, e.g. regarding libunwind

dra27 commented:
> Meanwhile: did you read the text written to introduce this PR? I think
> it may respond to a few of your comments, e.g. regarding libunwind

Ah - my bad! My suggestion for `-target-bindir` as `TARGET_BINDIR` should deal with that one completely - possibly with a minor change needed to `stdlib/Makefile` which I don't think should define it. I think with my suggestion for `--with-libunwind` and three environment variables that the code can be pretty much transplanted from the old configure script - happy to help.

shindere commented:
Also, you may want to go through the discussion on caml-devl about how
libunwind is supported. I really think that should be changed and I
think @damiendoligez agrees.

In any case, could that and the libunwind stuff be implemented in other
PRs that would be merged shortly after this one?

dra27 commented:
@shindere - I have. The question from caml-devel is whether macOS should ignore those settings when passed, but it still looks as though it should be possible to pass the include and lib directories separately - as it stands, there aren't enough options corresponding to the previously available (and used) ones.

I'm completely happy for the logic to be improved after autoconf is in, but the *interface* needs to be correct before it goes in - i.e. this PR must determine the final names of parameters and environment variables. That's what my review's attempted to address, hopefully without meandering to other things!

shindere commented:
David Allsopp (2018/12/09 05:53 -0800):
> @shindere - I have. The question from caml-devel is whether macOS
> should ignore those settings when passed, but it still looks as though
> it should be possible to pass the include and lib directories
> separately - as it stands, there aren't enough options corresponding
> to the previously available (and used) ones.

Yep I know. Since I wasn't sure I simply didn't continue on that, there
was already a lot to deal with.

> I'm completely happy for the logic to be improved after autoconf is
> in, but the *interface* needs to be correct before it goes in - i.e.
> this PR must determine the final names of parameters and environment
> variables. That's what my review's attempted to address, hopefully
> without meandering to other things!

Okay no problem, that sounds fine to me.

dra27 commented:
@shindere - please pull your GitHub branch before next force pushing... your last rebase blew away 1d6f270 (a `.gitattributes` update)... I've pushed it back!

dra27 commented:
@shindere - I just pushed e742127. Although the OCaml Travis job hasn't started yet, your local fork appears to be working with that shim. The default for `ASPP` with autoconf is picking up `$toolpref` which isn't strictly correct - I think we should either be reusing `CC` or redoing the test with `AC_CHECK_TOOL` (this would look for `i386-pc-linux-gnu-gcc` and on not finding it fall back to `gcc`). However, I think that can wait...

dra27 commented:
I just put your branch through precheck, [job 169](https://ci.inria.fr/ocaml/job/precheck/169/)

dra27 commented:
Assuming this passes precheck/Travis/AppVeyor, there are still the remaining questions in my review above - not hassling, just not leaving anyone tempted to merge this the second it passes CI 🙂

damiendoligez commented:
>just not leaving anyone tempted to merge this the second it passes CI

Who, me?

shindere commented:
David Allsopp (2018/12/09 05:12 -0800):
> I've gone through the mapping of options to the old `configure` script
> and where they appear in the new one. Reading the various guides for
> autoconf, it seems to me that `AC_ARG_WITH` is intended to be used for
> external dependency related things only and so should only relate to
> the libunwind and afl options. That seems a more important rule to
> regard than the nonsense in the autoconf manual about `AC_ARG_ENABLE`
> not altering any features (since they don't provide any other kind of
> parameter for things like flambda, which is a feature which alters
> other features).

Yeah I was aware of the distinction. I have to say I was not found of
using enable for features whose value could be something else than yes
or no, that's why I have chosen to use AC_ARG_WITH in places where it is
perhaps not conform to conventions. Is this something that absolutely
needs to be changed for this PR to get merged?

> The old `configure` made no use of environment variables, so there are
> a few arguments which really should change to `AC_ARG_VAR` now,
> especially when they relate to passing directories in. Users have been
> used to saying `./configure -cc cc` so hopefully will switch to the
> correct `./configure CC=cc` and not incorrect `CC=cc ./configure`.

Could you be more specific here and say which arguments you are talking
about, please?

> - The old `-lib*` parameter has no equivalent in the new one - should
> `$LIBS` be being put into `$cclibs`?

I think so, yes. Does that look okay?

> - This is more of a todo for me, but a reminder - a patch of mine
> suggested changing `PARTIALLD` to `PACKLD` in `configure`. This is a
> mistake which I'll change - `configure` should accept `PARTIALLD`
> but `Makefile.config` should only contain `PACKLD`.

So that has now been integrated to this PR, right?

> - The `-dldefs` parameter has no equivalent - as with `LIBS`,
> - shouldn't `CFLAGS` be read at some point?

Frankly, I was unsure what to do with that one. Is it really used
somewhere?

> - `--with-cplugins` should be added, solely to maintain the error
> message.

AIUI, by default an unrecognized option would raise a warning like:

```
configure: WARNING: unrecognized options: --with-cplugins
```

which will appear at the very bottom of configure's output, so not lost
in the middle of the "checking" messages.

Moreover, configure now has the `--enable-option-checking` which turns
warnings like the one shown above into errors.

Also, since, as you mentionned, every option has changed, people will
have to change their invocations to configure manually so if your
documentation about the arguments gets in, they will read it and see
that this option is simply no longer supported.

Given all this, I'm wondering whether this option really requires an
explicit handling?

@xavierleroy and @damiendoligez, what do you think?

> Every single argument has changed, apart from `--prefix`, `--bindir`,
> `--libdir`, `--mandir`, `--host`, `--target` and `--verbose`.

Well there's actually a syntactic change even for the
directory-related arguments, becuase the handwritten configure didn't
support =, i.e. one had to write
`--prefix /foo/bar`, which now becomes `--prefix=/foo/bar`.

> In doing this review, I now have a document which maps the old options
> to what should be their new equivalents. I think we should either
> insert a one-time script in front of `configure.ac` which translates
> the parameters (and which would never be updated with any new ones,
> obviously) or update the documentation with this mapping. My
> preference is for the script, which I'm happy to write, since it
> maintains as smooth a path as possible.

Discussed with @xavierleroy and @damiendoligez. The three of us would
prefer the documentation to the script. I, too, offered to provide that
documentation when I introduced the PR, jsut that I didn't know where it
should go and that's why it was not part of the initial submission.

> I don't think `--disable-shared` is working properly, but it's awkward
> testing both this and #2160 simultaneously, so I'll come back to that
> after this GPR is merged!

Okay let's do that, sorry for not having been careful enough.

> > +
> +AC_ARG_WITH([dllibs],
> +  [AS_HELP_STRING([--with-dllibs=LIBS],
> +    [use LIBS (in addition to dl) to load dynamic libs @<:@default=""@:>@])],
> +  [otherdllibs=$withval],
> +  [otherdllibs=""])
> +
> +AC_ARG_ENABLE([instrumented-runtime],
> +  [AS_HELP_STRING([--enable-instrumented-runtime],
> +    [build the instrumented runtime @<:@default=auto@:>@])],
> +  [],
> +  [enable_instrumented_runtime=auto])
> +
> +AC_ARG_ENABLE([pthread],
> +  [AS_HELP_STRING([--disable-pthread],
> +    [disable the POSIX thread library])])
> 
> I think this should be named `systhreads`, since the Windows implementation has nothing to do with POSIX:
> ```
> AC_ARG_ENABLE([systhreads-lib], 
>   [AS_HELP_STRING([--disable-systhreads-lib], 
>     [disable the native threads (POSIX/Win32) library])])
> ```

Any opinion @damiendoligez nd @xavierleroy?

> > +  [AS_HELP_STRING([--enable-instrumented-runtime],
> +    [build the instrumented runtime @<:@default=auto@:>@])],
> +  [],
> +  [enable_instrumented_runtime=auto])
> +
> +AC_ARG_ENABLE([pthread],
> +  [AS_HELP_STRING([--disable-pthread],
> +    [disable the POSIX thread library])])
> +
> +AC_ARG_WITH([libunwind],
> +  [AS_HELP_STRING([--with-libunwind],
> +    [location of the libunwind package])])
> +
> +AC_ARG_WITH([libunwind-include-dir],
> +  [AS_HELP_STRING([--with-libunwind-include-dir],
> +    [location of header files for libunwind])])
> 
> The unwind stuff seems to have several problems - there are large
> parts of the old `configure` (especially with macOS options) which
> don't appear to have been translated - is that intentional?

Yes, see the initial explanation of this GPR. Same goes for the support
of target-bindir.

> Most importantly, the default seems to have changed - libunwind is meant to
> be required unless it's explicitly disabled? I know there was a
> discussion on caml-devel, but I wonder if the proper interface for
> this should be `--with-libunwind` with `LIBUNWIND_PREFIX` (default
> `""`), `LIBUNWIND_INCLUDE` (default is either `""` or
> `$LIBUNWIND_PREFIX/include` if `LIBUNWIND_PREFIX` is defined) and
> `LIBUNWIND_LIB` (same system for the default)

If possible, I'd really like to postpone as much things as possible so
that each aspect can be delt with in an independent and smaller GPR.

> > +
> +AC_ARG_ENABLE([frame-pointers],
> +  [AS_HELP_STRING([--enable-frame-pointers],
> +    [use frame pointers in runtime and generated code])])
> +
> +AC_ARG_ENABLE([naked-pointers],
> +  [AS_HELP_STRING([--disable-naked-pointers],
> +    [do not allow naked pointers])])
> +
> +AC_ARG_ENABLE([spacetime],
> +  [AS_HELP_STRING([--enable-spacetime],
> +    [build the spacetime profiler])])
> +
> +AC_ARG_ENABLE([call-counts],
> +  [AS_HELP_STRING([--disable-call-counts],
> +    [disable the call counts in spacetime])])
> 
> `enable-call-counts` only makes sense with `--enable-spacetime`. I
> therefore wonder if it would be better to have `--spacetime=mode`
> where `mode` is `with-call-counts` (the default) or
> `without-call-counts`

Well, `--spacetime=mode` is not permitted, as is. We could have
`--with-spacetime=mode` or a`--enable-spacetime=mode`, though.


> > +    [build the spacetime profiler])])
> +
> +AC_ARG_ENABLE([call-counts],
> +  [AS_HELP_STRING([--disable-call-counts],
> +    [disable the call counts in spacetime])])
> +
> +AC_ARG_ENABLE([cfi],
> +  [AS_HELP_STRING([--disable-cfi],
> +    [disable the CFI directives in assembly files])])
> +
> +AC_ARG_ENABLE([installing-source-artifacts],
> +  [AS_HELP_STRING([--enable-installing-source-artifacts],
> +    [install *.cmt* and *.mli files])])
> +AC_ARG_ENABLE([installing-bytecode-programs],
> +  [AS_HELP_STRING([--enable-installing-bytecode-programs],
> +    [also install the bytecode versions of programs])])
> 
> As a matter of personal taste, I find these options unusually named.
> Can I propose `--enable-source-install` and
> `--enable-bytecode-install` - it's the `installing` which seems very
> strange.

I agree that the original names are not very nice. To be honest,
although your suggestions do look nicer, I tend to find them a bit
vague, not describing what exactly will be installed. But I have no
strong opinion here. Any other suggestion?


> 
> > +    [disable the call counts in spacetime])])
> +
> +AC_ARG_ENABLE([cfi],
> +  [AS_HELP_STRING([--disable-cfi],
> +    [disable the CFI directives in assembly files])])
> +
> +AC_ARG_ENABLE([installing-source-artifacts],
> +  [AS_HELP_STRING([--enable-installing-source-artifacts],
> +    [install *.cmt* and *.mli files])])
> +AC_ARG_ENABLE([installing-bytecode-programs],
> +  [AS_HELP_STRING([--enable-installing-bytecode-programs],
> +    [also install the bytecode versions of programs])])
> +
> +AC_ARG_ENABLE([native-compiler],
> +  [AS_HELP_STRING([--disable-native-compiler],
> +    [do not build the native compiler])])
> 
> Given again that it's changing name, can we consider
> `--enable-ocamlopt`?

Why not! Again it would be helpful to get other opinions, here.


> 
> > +    [install *.cmt* and *.mli files])])
> +AC_ARG_ENABLE([installing-bytecode-programs],
> +  [AS_HELP_STRING([--enable-installing-bytecode-programs],
> +    [also install the bytecode versions of programs])])
> +
> +AC_ARG_ENABLE([native-compiler],
> +  [AS_HELP_STRING([--disable-native-compiler],
> +    [do not build the native compiler])])
> +
> +AC_ARG_ENABLE([flambda],
> +  [AS_HELP_STRING([--enable-flambda],
> +    [enable flmabda optimizations])])
> +
> +AC_ARG_ENABLE([flambda-invariants],
> +  [AS_HELP_STRING([--enable-flambda-invariants],
> +    [enable invariants checks in flambda])])
> 
> As with `--enable-spacetime`, this is related to its parent option, so
> would `--flambda=with-invariants` be more logical?

Same response as above, one would have to use a --wit- of --enable-
prefix.


> 
> > +
> +AC_ARG_ENABLE([native-compiler],
> +  [AS_HELP_STRING([--disable-native-compiler],
> +    [do not build the native compiler])])
> +
> +AC_ARG_ENABLE([flambda],
> +  [AS_HELP_STRING([--enable-flambda],
> +    [enable flmabda optimizations])])
> +
> +AC_ARG_ENABLE([flambda-invariants],
> +  [AS_HELP_STRING([--enable-flambda-invariants],
> +    [enable invariants checks in flambda])])
> +
> +AC_ARG_WITH([target-bindir],
> +  [AS_HELP_STRING([--with-target-bindir],
> +    [location of binary programs on target system])])
> 
> This parameter isn't propagated anywhere - it should be in
> `Makefile.config.in`. However, given how it's used, it should really
> be `AC_ARG_VAR([TARGET_BINDIR], ...)`

As I said, I intended to fix this later, i.e. after the merge, during
the feature freeze.

> > +
> +AC_ARG_ENABLE([flambda],
> +  [AS_HELP_STRING([--enable-flambda],
> +    [enable flmabda optimizations])])
> +
> +AC_ARG_ENABLE([flambda-invariants],
> +  [AS_HELP_STRING([--enable-flambda-invariants],
> +    [enable invariants checks in flambda])])
> +
> +AC_ARG_WITH([target-bindir],
> +  [AS_HELP_STRING([--with-target-bindir],
> +    [location of binary programs on target system])])
> +
> +AC_ARG_WITH([reserved-header-bits],
> +  [AS_HELP_STRING([--with-reserved-header-bits=BITS],
> +    [reserve BITS bits in block headers for profiling info])])
> 
> This option hasn't quite preserved the semantics of the old one,
> although I'm not sure how relevant it is. Specifying
> `--with-reserved-header-bits` is also supposed to enable profinfo
> (even without spacetime). I wonder if it would be better to have
> `AC_ARG_ENABLE([profinfo], ...)` and have this as its parameter?

No opinion either, to be discussed with others, I think



> 
> > +# setting will be chosen by default, if no compile-time option is
> +# explicitly passed.
> +#
> +# The configure-time behavior of OCaml 4.05 and older was equivalent
> +# to --disable-force-safe-string --with-default-string=unsafe. OCaml 4.06
> +# and later uses --disable-force-safe-string --with-default-string=safe. We
> +# expect --enable-force-safe-string to become the default in the future.
> +
> +AC_ARG_ENABLE([force-safe-string],
> +  [AS_HELP_STRING([--disable-force-safe-string],
> +    [do not force strings to be safe])])
> +
> +AC_ARG_WITH([default-string],
> +  [AS_HELP_STRING([--with-default-string=mode],
> +    [whether strings should be safe or unsafe by default
> +      @<:@default=safe@:>@])])
> 
> This was a problem in the previous configure script, so perhaps now is
> the time to fix this properly. `--enable-force-safe-string
> --with-default-string=unsafe` is a conflicting configuration (I think
> it builds a compiler which displays an error message by default, but I
> haven't checked). I think this should one option
> `--enable-safe-string=mode` where `mode` is `no` (equivalent to
> `--disable-force-safe-string --with-default-string=unsafe`),
> `yes`/`default` (equivalent to `--disable-force-safe-string
> --with-default-string=safe`) or `always` (equivalent to
> `--enable-force-safe-string --with-default-string=safe`)

That seems easy to implement, if people agree upon this?


> 
> > +
> +AC_ARG_ENABLE([debug-runtime],
> +  [AS_HELP_STRING([--disable-debug-runtime],
> +    [do not build runtime with debugging support])])
> +
> +AC_ARG_ENABLE([debugger],
> +  [AS_HELP_STRING([--enable-debugger],
> +    [build the debugger @<:@default=yes@:>@])],
> +  [],
> +  [enable_debugger=yes])
> +
> +AC_ARG_WITH([dllibs],
> +  [AS_HELP_STRING([--with-dllibs=LIBS],
> +    [use LIBS (in addition to dl) to load dynamic libs @<:@default=""@:>@])],
> +  [otherdllibs=$withval],
> +  [otherdllibs=""])
> 
> This should be `AC_ARG_VAR([DLLIBS], ...)`

Sure, I can implement it that way.

shindere commented:
David Allsopp (2018/12/09 05:41 -0800):
> > Meanwhile: did you read the text written to introduce this PR? I think
> > it may respond to a few of your comments, e.g. regarding libunwind
> 
> Ah - my bad! My suggestion for `-target-bindir` as `TARGET_BINDIR`
> should deal with that one completely - possibly with a minor change
> needed to `stdlib/Makefile` which I don't think should define it. I
> think with my suggestion for `--with-libunwind` and three environment
> variables that the code can be pretty much transplanted from the old
> configure script - happy to help.

Okay will look into this, thanks. Can we agree to deal with this after
the merge, in one or two more specific GPRs?

shindere commented:
David Allsopp (2018/12/17 15:58 +0000):
> @shindere - I just pushed e742127. Although the OCaml Travis job
> hasn't started yet, your local fork appears to be working with that
> shim.

Thanks for the fix!

> The default for `ASPP` with autoconf is picking up `$toolpref`
> which isn't strictly correct - I think we should either be reusing
> `CC` or redoing the test with `AC_CHECK_TOOL` (this would look for
> `i386-pc-linux-gnu-gcc` and on not finding it fall back to `gcc`).
> However, I think that can wait...

Okay thanks, let's postpone that too, then.

shindere commented:
David Allsopp (2018/12/17 08:00 -0800):
> I just put your branch through precheck, [job
> 169](https://ci.inria.fr/ocaml/job/precheck/169/)

Would you mind telling me the results, please?

It's quite tedious to examine the results in Jenkins. When I do the
precheck myself I get e-mails, which is much easier for me to deal with.

dra27 commented:
@shindere:
 - precheck was basically a pass. Cygwin64 is broken (expected) and the PowerPC jobs fail because they need the configure argument updated, but these are low priority and almost certainly fine anyway
 - The Travis failure is fixed in #2209

I don't particularly agree with the urgency to merge this with known things not yet settled (especially names of new parameters), rather than working on it in this branch. However, on the basis that it's definitely working on all the tier 1 platforms, and the Windows builds are equivalently ported, I'm happy for this to go ahead, assuming everyone else is too, and we'll add more GPRs to it to pick up the remaining issues in the next few days.

The discussions on the changes to argument names and so forth really should be sorted before beta1, though - it'll be very weird changing those between two beta releases.

shindere commented:
Damien Doligez (2018/12/18 05:07 -0800):
> damiendoligez commented on this pull request.
> 
> 
> 
> > +# setting will be chosen by default, if no compile-time option is
> +# explicitly passed.
> +#
> +# The configure-time behavior of OCaml 4.05 and older was equivalent
> +# to --disable-force-safe-string --with-default-string=unsafe. OCaml 4.06
> +# and later uses --disable-force-safe-string --with-default-string=safe. We
> +# expect --enable-force-safe-string to become the default in the future.
> +
> +AC_ARG_ENABLE([force-safe-string],
> +  [AS_HELP_STRING([--disable-force-safe-string],
> +    [do not force strings to be safe])])
> +
> +AC_ARG_WITH([default-string],
> +  [AS_HELP_STRING([--with-default-string=mode],
> +    [whether strings should be safe or unsafe by default
> +      @<:@default=safe@:>@])])
> 
> In trunk, `-force-safe-string` takes precedence. (see lines 167-170 of `utils/clflags.ml`)
> 

Damien Doligez (2018/12/18 05:09 -0800):
> damiendoligez commented on this pull request.
> 
> 
> 
> > +# setting will be chosen by default, if no compile-time option is
> +# explicitly passed.
> +#
> +# The configure-time behavior of OCaml 4.05 and older was equivalent
> +# to --disable-force-safe-string --with-default-string=unsafe. OCaml 4.06
> +# and later uses --disable-force-safe-string --with-default-string=safe. We
> +# expect --enable-force-safe-string to become the default in the future.
> +
> +AC_ARG_ENABLE([force-safe-string],
> +  [AS_HELP_STRING([--disable-force-safe-string],
> +    [do not force strings to be safe])])
> +
> +AC_ARG_WITH([default-string],
> +  [AS_HELP_STRING([--with-default-string=mode],
> +    [whether strings should be safe or unsafe by default
> +      @<:@default=safe@:>@])])
> 
> I'd use `unsafe / safe / force` instead of `no / yes / always`.

@damiendoligez what do you think about @dra27's proposal? Shall we go
for that?

shindere commented:
David Allsopp (2018/12/19 16:08 +0000):
> @shindere:
>  - precheck was basically a pass. Cygwin64 is broken (expected) and
>  the PowerPC jobs fail because they need the configure argument
>  updated, but these are low priority and almost certainly fine anyway

Okay, thanks!

>  - The Travis failure is fixed in #2209

Right.

> I don't particularly agree with the urgency to merge this with known
> things not yet settled (especially names of new parameters), rather
> than working on it in this branch.

Well in my understanding, there are two very distinct things, here.

(1) Once this gets merged it willb be possible to do the feature freeze
which may be of some importance to @damiendoligez, but I'll let him
comment on that one.

(2) To be honest, working on this since last July (not full time, of
course), I start to get tired with this big GPR and that's why I'm
pushing for it to be merged, although it's very far from being perfect,
to get it away and be able to deal with smaller objects.

That being said I realise it's kind of irrational and I don't mind
spending more time tweaking the interface if @damiendoligez wishes so.

> However, on the basis that it's
> definitely working on all the tier 1 platforms, and the Windows builds
> are equivalently ported, I'm happy for this to go ahead, assuming
> everyone else is too, and we'll add more GPRs to it to pick up the
> remaining issues in the next few days.

I'm fine either way. Slight preference for an immediate merge but maybe
it's better not to follow me on this, I'll leave it up to
@damiendoliggez.

> The discussions on the changes to argument names and so forth really
> should be sorted before beta1, though - it'll be very weird changing
> those between two beta releases.

Fully agreed.

dra27 commented:
All good - but I don't think that merging this GPR immediately allows the freeze, inasmuch as some of the changes would be breaking (e.g. changing some things to `AC_ARG_VAR` and renaming some options mentioned above). But if it's easier to be looking at those in a fresh smaller GPR then that's fine!

We are also blocked on #2041, which I will start rebasing tomorrow.

dra27 commented:
@shindere - shall we give this one last rebase as that should then sort out Travis. We could also reconfigure the Jenkins workers and give it one final test in precheck? I'm happy to do the rebase and reconfigure, if you wish?

shindere commented:
@dra27 I'm fine doing the final steps myself, thanks!
I'll also celan up the history, actually already started to do so.

Will write a comment here when ready.

ksromanov commented:
> Well that line comes from `config/auto-aux/cckind.c`, see line 28. You may be right but (1) it's hard to say without being able to test with the xlc compiler 

I checked on AIX/xlc - both versions `#elif defined(__xlc__) && (__xlC__)` and `#elif defined(__xlc__) && defined(__xlC__)` work. Of course, the second looks better, so please correct in both places if you want.

shindere commented:
Just rebased on latest trunk and squashed all the commits into the main one.
Let me thank @dra27 again for his careful review and all his help.

This is currently going through precheck and hopefully through our other CI
solutions.

shindere commented:
Konstantin Romanov (2018/12/20 13:03 -0800):
> > Well that line comes from `config/auto-aux/cckind.c`, see line 28. You may be right but (1) it's hard to say without being able to test with the xlc compiler 
> 
> I checked on AIX/xlc - both versions `#elif defined(__xlc__) &&
> (__xlC__)` and `#elif defined(__xlc__) && defined(__xlC__)` work. Of
> course, the second looks better, so please correct in both places if
> you want.

The latest version of the GPR includes a fix, see `aclocal.m4:52`. What
do you mean by both places?

dra27 reviewed (approved):
Inria CI is a qualified (the failures which happened are either unrelated or now tested separately).

I propose merging once Travis/AppVeyor re-pass this!

shindere commented:
Thanks, David!

dra27 commented:
Spectrum Is Green! Here goes...

shindere commented:
David Allsopp (2018/12/21 07:02 -0800):
> Merged #2139 into trunk.

Wow. Thanks.

ksromanov commented:
> The latest version of the GPR includes a fix, see `aclocal.m4:52`. What do you mean by both places?

In `cckind.c` - https://github.com/ocaml/ocaml/blob/trunk/config/auto-aux/cckind.c#L28
But it is a cosmetic change.

shindere commented:
Konstantin Romanov (2018/12/21 07:19 -0800):
> In `cckind.c` - https://github.com/ocaml/ocaml/blob/trunk/config/auto-aux/cckind.c#L28
> But it is a cosmetic change.

Well now that the PR has been merged this file is not going to be used
any longer and will actually be removed soon so I think it's okay not to
fix it.

shindere commented:
@dra27, sorry for having such a bad memory: was it you who added

```
AC_CHECK_TOOL([DIRECT_CPP],[cpp])
AS_CASE([$host],
  [*-pc-windows], [CPP="$CC -nologo -EP"],
  [CPP=$DIRECT_CPP])
```

to configure.ac?

I am asking because this breaks the OCAML_CC_VENDOR macro, e.g.
even if configure is called with CC=clang, say, then the macro will call
gcc's preprocessor and then the C compiler vendor will be detected in a
wrong way. And same goes for xlc, for instance.

Any idea how this should be fixed?

ksromanov commented:
The breaking part is `default` choice of `AC_CASE`. Perhaps you can remove this snippet at all and add

```
      CPP="$CC -nologo -EP";
````

to the snippet couple lines above - https://github.com/ocaml/ocaml/blob/trunk/configure.ac#L395  
Of course, comment about `fcc -E` is very important, though I could not find, which tool is `fcc` (it is impossible to google it out). My guess - something from Visual Studio.

shindere commented:
Well isn't fcc just gcc with a typo? :)

dra27 commented:
@shindere - no, the correction I was made was to remove the hardcoding of `CPP=cpp` to `CPP=$DIRECT_CPP`, since that was breaking the mingw ports if Cygwin's gcc wasn't also installed (they need to use `i686-w64-mingw32-cpp` or `x86_64-w64-mingw32-cpp`, which is why I added the `AC_CHECK_TOOL` call)

shindere commented:
Gabriel Scherer (2019/04/15 04:46 -0700):
> There is a conceptual difference between:
> - `--force-safe-string`, which must be a configure-time option that changes the runtime definitions and has a global impact on all modules compiled and linked by the resulting compiler
> - `--default-string={safe,unsafe}` which is "merely" a convenience
> - feature to set a default for compile-time options, rather than a
> - configure-time option by nature (`afl` is the same, etc.). It does
> - nothing more than change the default for a choice that the user
> - makes on a per-invocation basis.

Sure, I agree. I'm just unclear what you conclude there from / which
consequences this has?

shindere commented:
Gabriel Scherer (2019/04/15 04:48 -0700):
> In my opinion this conceptual difference justifies *not* showing the
> three options as a single knob to fix at configure-time, but two
> separate kind of options. There is a dependency between then
> (`--force-safe-string` makes it incorrect to change the per-invocation
> value, so it makes `--default-string` useless and meaningless), but
> that is common across the configuration space, for example
> `--disable-native-compiler` will make many other settings meaningless.

OK I think I understand now what you meant, sorry.
FWIW, another reason would be that it was like this before so it would
be more conservative to keep two distinct options.
